### PR TITLE
feat(expr): implement greatest and least function

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Complete https://github.com/risingwavelabs/risingwave/issues/12761

The code of cmp unit test is refactored accordingly.

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support `greatest` and `least` pg function.

```
GREATEST(value [, ...])
```

```
LEAST(value [, ...])
```

The GREATEST and LEAST functions select the largest or smallest value from a list of any number of expressions. The expressions must all be convertible to a common data type, which will be the type of the result.

NULL values in the argument list are ignored. The result will be NULL only if all the expressions evaluate to NULL. (This is a deviation from the SQL standard. According to the standard, the return value is NULL if any argument is NULL. Some other databases behave this way.)

Refer to https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-GREATEST-LEAST

### CODE REVIEW

**Review by TennyZhuang — APPROVED** [review]
LGTM

File: src/frontend/src/binder/expr/function.rs [hunk 1]
```diff
@@ -672,6 +672,20 @@ impl Binder {
             )
         }
 
+        fn reduce(r#type: ExprType) -> Handle {
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/function.rs:675]: There is variadic function support: #12178

> The expressions must all be convertible to a common data type, which will be the type of the result (see [Section 10.5](https://www.postgresql.org/docs/current/typeconv-union-case.html) for details).

In the frontend this can be achieved with [`align_types`](https://github.com/risingwavelabs/risingwave/blob/3579e74ea9194170fcec8f1f74c4f4aac6b4b2fd/src/frontend/src/expr/type_inference/func.rs#L319).

Comment (wenym1) [reply src/frontend/src/binder/expr/function.rs:675]: I just checked the function currently using the variadic input. The input type of the function implementations are all `f(row: impl Row)` like `array`
```
#[function("array(...) -> list")]
fn array(row: impl Row) -> ListValue {
    ListValue::new(row.iter().map(|d| d.to_owned_datum()).collect())
}
```

To implement the functions in this PR, if the input types are also `impl Row`, we will have to cast the type to the inner type manually, which is verbose and dirty. Is is possible to do things like the following with the proc macro? 
```
fn generic_greatest<T>(input: impl IntoIterator<Item=Option<T>>) -> Option<T> {}
```

If not I think the current implementation is better.

Comment (xiangjinwu) [reply src/frontend/src/binder/expr/function.rs:675]: > we will have to cast the type to the inner type manually, which is verbose and dirty

You meant `T::try_from`? Anyways it would be a backward compatible change from a binary tree to a flat one later. cc @wangrunji0408

Comment (wangrunji0408) [reply src/frontend/src/binder/expr/function.rs:675]: >  Is is possible to do things like the following with the proc macro?
> ```
> fn generic_greatest<T>(input: impl IntoIterator<Item=Option<T>>) -> Option<T> {}
> ```

Yes, it's possible. But it requires some effort to support this in the macro. If this does help, I may add this support in the near future.

> we will have to cast the type to the inner type manually, which is verbose and dirty

Can we compare the `Datum` directly without downcast? I think it should be correct if all inputs have been casted to the common type.

File: src/expr/impl/src/scalar/cmp.rs [hunk 2]
```diff
@@ -361,11 +416,34 @@ fn batch_is_not_null(a: &impl Array) -> BoolArray {
...
+    let data = (a.data() & a.null_bitmap()) | (b.data() & b.null_bitmap());
+    BoolArray::new(data, bitmap)
+}
+
+fn boolarray_least(a: &BoolArray, b: &BoolArray) -> BoolArray {
+    // Use `or` because null is ignored.
+    let bitmap = a.null_bitmap() | b.null_bitmap();
+    // Three cases that can be true
+    //   1. left side has data and is true but right side has no data
+    //   2. right side has data and is true but right side has no data
```
Comment (wangrunji0408) [inline src/expr/impl/src/scalar/cmp.rs:432]:
```suggestion
    //   2. right side has data and is true but left side has no data
```

File: src/frontend/src/binder/expr/function.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/frontend/src/binder/expr/function.rs:675]: Btw, it seems that implicit cast to the common type is needed in the frontend. Otherwise, the backend function doesn't know how to compare values between string and int.

Comment (xiangjinwu) [reply src/frontend/src/binder/expr/function.rs:675]: > Can we compare the Datum directly without downcast?

No because there are `null first` and `null last`. For the 2 functions here we actually pick neither but ignore nulls.

https://github.com/risingwavelabs/risingwave/blob/d2ec83c88b76e2a0d9b2c9ec59b05dd43edca780/src/common/src/types/mod.rs#L547-L551

Comment (wenym1) [reply src/frontend/src/binder/expr/function.rs]: I have implemented the variadic version with `f(row: impl Row)` and `T::try_from`. The implementation looks much better now. Thanks for the suggestions!

File: src/frontend/src/expr/type_inference/func.rs [hunk 3]
```diff
@@ -537,6 +548,10 @@ fn infer_type_for_special(
             ensure_arity!("vnode", 1 <= | inputs |);
             Ok(Some(DataType::Int16))
         }
+        ExprType::Greatest | ExprType::Least => {
+            ensure_arity!("greatest/least", | inputs | > 0);
```
Comment (xiangjinwu) [inline src/frontend/src/expr/type_inference/func.rs:552]: There is `ensure_arity!("greatest/least", 1 <= | inputs |);`

File: src/common/src/monitor/connection.rs [hunk 4]
```diff
@@ -226,7 +226,7 @@ where
     }
 
     fn call(&mut self, uri: Uri) -> Self::Future {
-        let endpoint = format!("{:?}:{:?}", uri.host(), uri.port());
+        let endpoint = format!("{:?}", uri.host());
```
Comment (xiangjinwu) [inline src/common/src/monitor/connection.rs:229]: Seems do not belong to this PR

File: src/expr/impl/src/scalar/cmp.rs [hunk 5]
```diff
@@ -287,6 +289,64 @@ fn is_not_null<T>(v: Option<T>) -> bool {
...
+pub fn general_variadic_greatest<T>(row: impl Row) -> Option<T>
+where
+    T: Scalar,
+    for<'a> <T as Scalar>::ScalarRefType<'a>: TryFrom<ScalarRefImpl<'a>> + Ord + Debug,
+{
+    row.iter()
+        .flatten()
+        .map(
+            |scalar| match <<T as Scalar>::ScalarRefType<'_>>::try_from(scalar) {
+                Ok(v) => v.to_owned_scalar(),
```
Comment (xiangjinwu) [inline src/expr/impl/src/scalar/cmp.rs:314]: It seems we can `max` on `ScalarRef` and only `to_owned_scalar` on the max found at the end.

**Review by xiangjinwu — APPROVED** [review]
Rest lgtm.

Feel free to update the unit test expectation on ambiguous signatures to make it pass. It is a false alarm that `greatest()` requires special return-type inference logic:
* `greatest()` is allowed by the variadic macro, but we have disallowed it in frontend with `ensure_arity` already.
* `greatest(..)` does have its special return-type inference logic (`align_types`).

File: src/frontend/src/binder/expr/function.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/frontend/src/binder/expr/function.rs]: Just fyi I found a case where `align_type` ([pg](https://www.postgresql.org/docs/16/typeconv-union-case.html)) is required and a binary expr tree would fail:
```
dev=> select greatest('12', '2');
 greatest 
----------
 2
(1 row)

dev=> select greatest('12', '2', 0);
 greatest 
----------
       12
(1 row)
```
The former fallback to `varchar`/`text` but the latter infers all 3 to be `int`. It is not equivalent to `greatest(greatest('12', '2'), 0)`.

**Review by hzxa21 — APPROVED** [review]
LGTM.

Need to update the release notes as well.

**Comment by xiangjinwu** [comment]
> You missed e2e tests for the streaming mode.

These 2 are not aggregate or set returning functions. For pure scalar functions, we do not repeat the test in different configurations.
